### PR TITLE
feat: add source-type and cluster-id flag support for kcp scan self-managed-connectors

### PR DESCRIPTION
**Pull request by pinaruyar**
## Summary                                                                                                                              
  - **Removed `--cluster-arn` flag** - replaced with single `--cluster-id` that works for both MSK and OSK                                
  - Add optional `--source-type` flag to `self-managed-connectors` command                                                                
  - Maintain auto-detection as default (ARN = MSK, non-ARN = OSK)                                                                         
  - Add cluster identifier helper for MSK and OSK sources                                                                                 
  - Add comprehensive test coverage including integration test environment                                                                
                                                                                                                                          
  ## Changes                                                                                                                              
                                                                                                                                          
  **Core Functionality:**                                                                                                                 
  - **Removed separate `--cluster-arn` flag** - now uses unified `--cluster-id` for both MSK ARNs and OSK cluster IDs
  - Added `--source-type` optional flag (`msk` | `osk`) with validation                                                                   
  - Auto-detection: if `--source-type` not specified, detects from `--cluster-id` format                                                  
    - Starts with `arn:` → MSK                                                                                                            
    - Non-ARN → OSK                                                                                                                       
  - Explicit `--source-type` overrides auto-detection (handles edge cases like OSK cluster IDs starting with "arn:")                      
  - New `cluster_identifier.go` and tests for detecting cluster types                                                                     
  - Updated `self_managed_connectors_scanner.go` to support source-type parameter                                                         
  - Updated `cmd_scan_self_managed_connectors.go` with simplified single `--cluster-id` flag and auto-detection logic                     
                                                                                                                                          
  **Before (separate flags):**                                                                                                            
  ```bash                                                                                                                                 
  # MSK required --cluster-arn                                                                                                            
  --cluster-arn arn:aws:kafka:...                                                                                                         
                                                                                                                                          
  # OSK required --cluster-id                                                                                                             
  --cluster-id my-cluster                                                                                                                 
                                                                                                                                          
  After (unified flag):                                                                                                                   
  # Both use --cluster-id (auto-detects MSK vs OSK)                                                                                       
  --cluster-id arn:aws:kafka:...        # MSK      
  --cluster-id my-cluster               # OSK                                                                                             
  --cluster-id any-id --source-type osk # Override
                                                                                                                                          
  Test Coverage:  
  - 26/26 unit tests pass (4 new tests added)                                                                                             
  - New tests:                                                                                                                            
    - TestParseScanSelfManagedConnectorsOpts_ExplicitSourceType_MSK
    - TestParseScanSelfManagedConnectorsOpts_ExplicitSourceType_OSK                                                                       
    - TestParseScanSelfManagedConnectorsOpts_InvalidSourceType                                                                            
    - TestParseScanSelfManagedConnectorsOpts_ExplicitSourceType_OverridesAutoDetection                                                    
  - Added Kafka Connect service to OSK integration test environment (docker-compose.yml)                                                  
    - Enables end-to-end testing of connector scanning                                                                                    
    - Exposes Connect REST API on port 8083                                                                                               
    - Auto-creates test connector on startup                                                                                              
                                                                                                                                          
  Usage Examples                                                                                                                          
                                                                                                                                          
  Auto-detection (recommended):                                                                                                           
  # MSK cluster (detected from ARN)
  kcp scan self-managed-connectors \
    --state-file kcp-state.json \                                                                                                         
    --connect-rest-url http://connect:8083 \                                                                                              
    --cluster-id arn:aws:kafka:us-east-1:123456789012:cluster/my-cluster/abc-123 \                                                        
    --use-unauthenticated                                                                                                                 
                                                                                                                                          
  # OSK cluster (detected from non-ARN format)                                                                                            
  kcp scan self-managed-connectors \                                                                                                      
    --state-file kcp-state.json \                                                                                                         
    --connect-rest-url https://connect.example.com:8083 \                                                                                 
    --cluster-id production-kafka \                      
    --use-sasl-scram \                                                                                                                    
    --sasl-scram-username admin \
    --sasl-scram-password secret                                                                                                          
                  
  Explicit override:                                                                                                                      
  # Force OSK even if cluster-id looks like an ARN
  kcp scan self-managed-connectors \                                                                                                      
    --state-file kcp-state.json \   
    --connect-rest-url http://connect:8083 \
    --cluster-id arn:custom:id \
    --source-type osk \                                                                                                                   
    --use-unauthenticated
                                                                                                                                          

  ✅ Tested locally with OSK cluster and MSK playground
  - Auto-detection correctly identifies OSK from non-ARN cluster-id
  - Explicit --source-type override works                                                                                                 
  - Connector data saved to state file successfully
                                                                                                                                          
  Reviewer Notes                                                                                                                          
   
  - Breaking change: Removes --cluster-arn flag in favor of unified --cluster-id (simplifies UX)                                          
  - Backward compatible: Auto-detection maintains existing behavior for users who provide ARNs vs non-ARN IDs
  - Edge case handling: Explicit source-type allows override if OSK cluster IDs happen to start with "arn:"                               
  - Test infrastructure: Kafka Connect added to OSK test environment enables future connector-related testing                             
  - Help text: Updated to show three examples (MSK auto-detect, OSK auto-detect, explicit override)                                       
                                                                                                                                          
  🤖 Generated with https://claude.com/claude-code

### CODE REVIEW

**Comment by pinaruyar** [comment]
Testing: scan self-managed-connectors Command                                                                                                                                                                        
                                                                                                                                                   
  Test Environment Setup                                                                                                                           
                                                                                                                                                   
  Prerequisites                                                                                                                                    
                  
  - Docker Desktop running                                                                                                                         
  - Go 1.24+
  - kcp binary built                                                                                                                               
                                                                                                                                                   
  Step 1: Start OSK Test Environment                                                                                                               
                                                                                                                                                   
  Navigate to the OSK integration test directory:                                                                                                  
  cd ~/repos/kcp/integration-tests/osk-scan
                                                                                                                                                   
  Start Kafka and Kafka Connect containers:
  docker-compose up -d kafka kafka-connect

  Wait ~30 seconds for services to start.                                                                                                          
   
  Step 2: Verify Services Are Running                                                                                                              
                  
  Check container status:                                                                                                                          
  docker-compose ps
                                                                                                                                                   
  Expected output:
  NAME                         IMAGE                                 COMMAND                   SERVICE         STATUS
  kcp-test-osk-kafka           confluentinc/cp-kafka:7.6.0           ...                       kafka           Up    
  kcp-test-osk-kafka-connect   confluentinc/cp-kafka-connect:7.6.0   ...                       kafka-connect   Up                                  
                                                                                                                                                   
  Verify Kafka Connect REST API:                                                                                                                   
  curl http://localhost:8083/                                                                                                                      
                  
  Expected output:                                                                                                                                 
  {"version":"7.6.0-ccs","commit":"...","kafka_cluster_id":"UnIfIeDtEsTcLuStErId0w"}

  Step 3: Create Test Connector                                                                                                                    
   
  Create a MirrorMaker 2 connector for testing:                                                                                                    
  curl -X POST http://localhost:8083/connectors -H "Content-Type: application/json" -d '{
    "name": "test-mirror-source",                                                                                                                  
    "config": {                                                                                                                                    
      "connector.class": "org.apache.kafka.connect.mirror.MirrorSourceConnector",                                                                  
      "tasks.max": "1",                                                                                                                            
      "source.cluster.alias": "source",                                                                                                            
      "target.cluster.alias": "target",                                                                                                            
      "source.cluster.bootstrap.servers": "localhost:9092",                                                                                        
      "target.cluster.bootstrap.servers": "localhost:9092",
      "topics": "test-topic-1"                                                                                                                     
    }             
  }'                                                                                                                                               
                  
  Verify connector exists:                                                                                                                         
  curl -s http://localhost:8083/connectors

  Expected output:                                                                                                                                 
  ["test-mirror-source"]
                                                                                                                                                   
  Step 4: Create OSK State File
                                                                                                                                                   
  Navigate back to kcp root:                                                                                                                       
  cd ~/repos/kcp                                                                                                                                   
                                                                                                                                                   
  Scan the OSK cluster to create a state file:
  ./kcp scan clusters \                                                                                                                            
    --source-type osk \
    --state-file test-osk-state.json \                                                                                                             
    --credentials-file integration-tests/osk-scan/credentials/kafka-plaintext.yaml
                                                                                                                                                   
  Expected output:                                                                                                                                 
  ✅ Scan completed successfully                                                                                                                   
     Scanned 1 cluster(s)                                                                                                                          
     State file: test-osk-state.json                                                                                                               
                                                                                                                                                   
  Verify state file contains the OSK cluster:
  cat test-osk-state.json | jq '.osk_sources.clusters[0].id'                                                                                       
                                                            
  Expected output:                                                                                                                                 
  "osk-kafka"                                                                                                                                      
                                                                                                                                                   
  Test Cases                                                                                                                                       
                  
  Test Case 1: Auto-Detection (No --source-type flag)                                                                                              
                  
  Command:                                                                                                                                         
  ./kcp scan self-managed-connectors \
    --state-file test-osk-state.json \                                                                                                             
    --connect-rest-url http://localhost:8083 \
    --cluster-id osk-kafka \
    --use-unauthenticated                                                                                                                          
   
  Expected Output:                                                                                                                                 
  🚀 Starting self-managed connector scan for cluster osk-kafka
    🔍 Found 1 connectors                                      
    ✅ Successfully retrieved connector details for 1 connectors                                                                                   
  ✅ Self-managed connector scan complete for cluster osk-kafka 
                                                                                                                                                   
  Verification:   
  cat test-osk-state.json | jq '.osk_sources.clusters[0].kafka_admin_client_information.self_managed_connectors'                                   
                  
  Should show:                                                                                                                                     
  {               
    "connectors": [                                                                                                                                
      {                                                                                                                                            
        "name": "test-mirror-source",                                                                                                              
        "config": {                                                                                                                                
          "connector.class": "org.apache.kafka.connect.mirror.MirrorSourceConnector",                                                              
          "name": "test-mirror-source",                                              
          "source.cluster.alias": "source",                                                                                                        
          "source.cluster.bootstrap.servers": "localhost:9092",                                                                                    
          "target.cluster.alias": "target",                                                                                                        
          "target.cluster.bootstrap.servers": "localhost:9092",                                                                                    
          "tasks.max": "1",                                                                                                                        
          "topics": "test-topic-1"                                                                                                                 
        }                                                                                                                                          
      }                                                                                                                                            
    ]             
  }                                                                                                                                                
                                                                                                                                                   
  Result: ✅ Auto-detection worked (detected OSK from non-ARN cluster-id)
                                                                                                                                                   
  ---             
  Test Case 2: Explicit --source-type osk                                                                                                          
                                                                                                                                                   
  Command:
  ./kcp scan self-managed-connectors \                                                                                                             
    --state-file test-osk-state.json \                                                                                                             
    --connect-rest-url http://localhost:8083 \
    --cluster-id osk-kafka \                                                                                                                       
    --source-type osk \     
    --use-unauthenticated                                                                                                                          
                         
  Expected Output:
  🚀 Starting self-managed connector scan for cluster osk-kafka                                                                                    
    🔍 Found 1 connectors                                      
    ✅ Successfully retrieved connector details for 1 connectors                                                                                   
  ✅ Self-managed connector scan complete for cluster osk-kafka 

  Result: ✅ Explicit source-type override worked                                                                                                  
   
  ---                                                                                                                                              
  Test Case 3: Verify Help Shows New Flag
                                                                                                                                                   
  Command:
  ./kcp scan self-managed-connectors --help                                                                                                        
                  
  Expected Output (excerpt):
  Optional Flags:
        --source-type string   Source type: 'msk' or 'osk'. If not specified, auto-detects from cluster-id format (ARN = MSK, non-ARN = OSK).
                                                                                                                                             
  Result: ✅ Help text shows optional source-type flag                                                                                             
                                                                                                                                                   
  ---                                                                                                                                              
  Test Case 4: Invalid source-type Value                                                                                                           
                  
  Command:
  ./kcp scan self-managed-connectors \
    --state-file test-osk-state.json \
    --connect-rest-url http://localhost:8083 \
    --cluster-id osk-kafka \                                                                                                                       
    --source-type invalid \                                                                                                                        
    --use-unauthenticated                                                                                                                          
                                                                                                                                                   
  Expected Output:
  Error: invalid source-type: invalid (must be 'msk' or 'osk')

  Result: ✅ Validation works (covered by unit tests)                                                                                              
   
                                                                                                                                      
  Cleanup                                                                                                                                               
  Stop Test Environment                                                                                                                            
   
  cd ~/repos/kcp/integration-tests/osk-scan                                                                                                        
  docker-compose down

  Remove Test State File

  cd ~/repos/kcp
  rm test-osk-state.json

  Notes                                                                                                                                             
  - The test environment uses unauthenticated Connect for simplicity                                                                               
  - For production MSK testing, you would need:
    - Valid MSK cluster ARN in state file                                                                                                          
    - Connect cluster URL (likely internal)
    - Appropriate authentication (SASL/SCRAM or TLS)
  - The connector status endpoint returns 404 initially because the connector tasks haven't fully started - this is expected and handled gracefully
   with a warning

File: cmd/scan/self_managed_connectors/cmd_scan_self_managed_connectors.go [hunk 1]
```diff
@@ -31,9 +32,32 @@ var (
...
+  # Explicitly specify source type (overrides auto-detection)
+  kcp scan self-managed-connectors \
+    --state-file kcp-state.json \
+    --connect-rest-url http://connect:8083 \
+    --cluster-id my-cluster \
+    --source-type osk \
+    --use-unauthenticated`,
 		SilenceErrors: true,
 		PreRunE:       preRunScanSelfManagedConnectors,
 		RunE:          runScanSelfManagedConnectors,
```
Comment (adrian-januzi) [inline cmd/scan/self_managed_connectors/cmd_scan_self_managed_connectors.go:63]: We probably want to unhide the command by removing line 64/setting `hidden: false`

File: cmd/scan/self_managed_connectors/cluster_identifier.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+		}
+		return utils.ExtractClusterNameFromArn(clusterArn)
+	case "osk":
+		if clusterID == "" {
+			return "unknown-cluster"
+		}
+		return clusterID
+	default:
+		return "unknown-cluster"
+	}
```
Comment (adrian-januzi) [inline cmd/scan/self_managed_connectors/cluster_identifier.go]: We should probably move this into `kcp/internal/utils` as a general function. At the moment based on where it is placed, it appears to be strictly a `self-managed-connectors` scoped function.

File: cmd/scan/self_managed_connectors/cmd_scan_self_managed_connectors.go [hunk 3]
```diff
@@ -156,16 +190,54 @@ func parseScanSelfManagedConnectorsOpts() (*SelfManagedConnectorsScannerOpts, er
...
-	if err != nil {
-		return nil, fmt.Errorf("cluster not found in state file: %v", err)
+	// Determine source type: use explicit flag if provided, otherwise auto-detect from cluster ID format
+	var detectedSourceType string
+	var clusterArn string
+	var oskClusterID string
+
+	if sourceType != "" {
+		// Validate explicit source type
+		if sourceType != "msk" && sourceType != "osk" {
```
Comment (adrian-januzi) [inline cmd/scan/self_managed_connectors/cmd_scan_self_managed_connectors.go:200]: Instead of `sourceType != ` a string (`"msk"`/`"osk"`), we should probably use the typed enums tucked away [here](https://github.com/confluentinc/kcp/blob/main/internal/types/state.go#L929-L935) instead:

```golang
types.SourcETypeMSK
types.SourceTypeOSK
```

File: cmd/scan/self_managed_connectors/cmd_scan_self_managed_connectors.go [hunk 4]
```diff
@@ -156,16 +190,54 @@ func parseScanSelfManagedConnectorsOpts() (*SelfManagedConnectorsScannerOpts, er
...
+			return nil, fmt.Errorf("invalid source-type: %s (must be 'msk' or 'osk')", sourceType)
+		}
+		detectedSourceType = sourceType
+	} else {
+		// Auto-detect from cluster ID format
+		if strings.HasPrefix(clusterID, "arn:") {
+			detectedSourceType = "msk"
+		} else {
+			detectedSourceType = "osk"
```
Comment (adrian-januzi) [inline cmd/scan/self_managed_connectors/cmd_scan_self_managed_connectors.go]: Same again, use `types.SourceType<MSK | OSK>`

File: integration-tests/osk-scan/docker-compose.yml [hunk 5]
```diff
@@ -103,6 +103,67 @@ services:
...
+        echo "Creating sample file source connector..."
+        curl -X POST http://localhost:8083/connectors -H "Content-Type: application/json" -d '{
+          "name": "test-file-source",
+          "config": {
+            "connector.class": "org.apache.kafka.connect.file.FileStreamSourceConnector",
+            "tasks.max": "1",
+            "file": "/tmp/test-input.txt",
+            "topic": "test-connect-topic"
+          }
+        }' || true
```
Comment (adrian-januzi) [inline integration-tests/osk-scan/docker-compose.yml:165]: nitpick: wouldn't this mean that if the test failed to deploy the connect cluster, we wouldn't know as the test would still pass due to the `|| true`?

**Comment by adrian-januzi** [comment]
lgtm:

```shell
[ec2-user@ip-10-0-10-118 ~]$ ./kcp scan self-managed-connectors     --state-file kcp-state.json     --connect-rest-url http://XX.XX.XX.XX:8083     --use-unauthenticated --cluster-id arn:aws:kafka:us-east-1:XXXXXXXX:cluster/kcp-playground/0251db9f-2675-4d50-b9f1-97e171c6b3ed-5

┌─────────────────────────────────────────────────────────────────────────┐
│ ⚠️  WARNING: This is a development build                                │
│ Official releases: https://github.com/confluentinc/kcp/releases         │
└─────────────────────────────────────────────────────────────────────────┘

Executing kcp with build version=0.0.0-localdev commit=6edd514116f83a1f42b5cfa42f0cc80857583c74 date=2026-04-28T14:20:14Z
🚀 Starting self-managed connector scan for cluster kcp-playground
  🔍 Found 1 connectors
  ✅ Successfully retrieved connector details for 1 connectors
✅ Self-managed connector scan complete for cluster kcp-playground
[ec2-user@ip-10-0-10-118 ~]$
[ec2-user@ip-10-0-10-118 ~]$
[ec2-user@ip-10-0-10-118 ~]$
[ec2-user@ip-10-0-10-118 ~]$ jq '.msk_sources.regions[].clusters[]
      | select(.name == "kcp-playground")
      | .kafka_admin_client_information.self_managed_connectors' \
    kcp-state.json
{
  "connectors": [
    {
      "name": "datagen-pageviews",
      "config": {
        "connector.class": "io.confluent.kafka.connect.datagen.DatagenConnector",
        "iterations": "10000000",
        "kafka.topic": "pageviews",
        "key.converter": "org.apache.kafka.connect.storage.StringConverter",
        "max.interval": "100",
        "name": "datagen-pageviews",
        "quickstart": "pageviews",
        "tasks.max": "1",
        "value.converter": "org.apache.kafka.connect.json.JsonConverter",
        "value.converter.schemas.enable": "false"
      },
      "state": "RUNNING"
    }
  ]
}
```

**Comment by adrian-januzi** [comment]
One more thing I've thought of now - you mentioned that the Connect integration tests take long and are opt-in for now. Would it be possible to add them to the `make` file so we can run them like `make test-kafka-connect`. Then also in the Semaphore pipeline add the following like [here](https://github.com/confluentinc/kcp/blob/main/.semaphore/semaphore.yml#L103-L106) so that the Connect integration tests only run when specific files are changed in a commit, something like this:

```yaml
    run:
      when: "change_in(['/cmd/scan/self_managed_connectors/', '/integration-tests/osk-scan/docker-compose.yml', '/integration-tests/osk-scan/run-connect.sh', '/Makefile', '/.semaphore/semaphore.yml'], {default_branch:
   'main'})"
```

Therefore, we don't need to manually worry about running them for regressions + they are easy to run through a simple make command.

---

One other thing is that I think we'd want to address [this](https://github.com/confluentinc/kcp/pull/280#discussion_r3155527418) just to ensure that we don't always have a green pipeline even if things failed.

File: integration-tests/osk-scan/docker-compose.yml (hunk 5 above)
Comment (pinaruyar) [reply integration-tests/osk-scan/docker-compose.yml:165]: this true command is related to the specific connector rather than kafka connect itself. And creating a connector is optional in this case (or it might already exist etc.) This ensures Docker container doesn't crash if creating this optional test connector fails in a test env. 
It is also a consistent thing going on this file. We might wanna discuss if we wanna keep followings:
_kafka-storage format || true 
kafka-configs  || true_

**Review by adrian-januzi — APPROVED** [review]
